text/0000-improved-binary-gem-support.md

```diff
@@ -0,0 +1,466 @@
+ ["rb30", "universal_darwin_20"],
+ ["rb30", "darwin"],
+ ["rb34", "any"],
+ ["rb3", "any"],
+ ["rb33", "any"],
+ ["rb32", "any"],
+ ["rb31", "any"],
+ ["rb30", "any"],
+ ["any", "arm64_darwin_24"],
+ ["any", "universal_darwin_24"],
+ ["any", "arm64_darwin"],
+ ["any", "universal_darwin"],
+ ["any", "arm64_darwin_23"],
+ ["any", "universal_darwin_23"],
+ ["any", "arm64_darwin_22"],
+ ["any", "universal_darwin_22"],
+ ["any", "arm64_darwin_21"],
+ ["any", "universal_darwin_21"],
+ ["any", "arm64_darwin_20"],
+ ["any", "universal_darwin_20"],
+ ["any", "darwin"],
+ ["any", "any"]]
+```
+
+# Reference-level explanation
+
+The implementation adds four new classes to `lib/rubygems/platform/`:
+
+**`Gem::Platform::GemBinary`** - Parses gem binary-format platform strings (`gb-{abi}-{platform}`)
+- Handles multi-tag specifications (e.g., `rb33.rb32` for multiple Ruby versions)
+- Provides matching logic against Ruby environments
+- Expands tag combinations for compatibility checking
+
+**`Gem::Platform::Specific`** - Represents the current Ruby environment with precision
+- Detects Ruby engine, version, and ABI version
+- Identifies Linux libc type and version through ELF analysis
+- Generates appropriate compatible gem binary tags for the environment
+
+**`Gem::Platform::ELFFile`** - Analyzes ELF binaries on Linux
+- Reads interpreter path to identify musl vs glibc
+- Uses `Etc::CS_GNU_LIBC_VERSION` or `ldd` to determine minimum glibc version
+- Uses the musl interpreter to determine minimum musl version
+- Provides platform detection without external dependencies
+
+**`Gem::Platform::GlibcLinux`/`Gem::Platform::MuslLinux`** - Linux compatibility standards
+- Implements a pared down equivalent of the PEP 600 (manylinux) and PEP 656 (musllinux) specifications
+- Maps glibc/musl versions to compatibility tags
+- Allows for compact representation of cross-distribution compatible binaries
+
+## Platform Resolution Changes
+
+The platform matching system now uses `Gem::Platform::Specific.local` instead of `Gem::Platform.local` for more precise matching. The `sort_priority` method in `Gem::Platform` now assigns:
+
+1. Ruby platform: priority -1 (highest, for pure Ruby gems)
+2. gem binary platforms: priority 2 (preferred for binary gems)
+3. Traditional platforms: priority 1 (fallback for legacy gems)
+
+This ensures matching gem binary-platformed gems are preferred when available while maintaining backward compatibility.
+
+# Drawbacks
+
+- **Increased complexity** - The resolver must handle more platform variants, though this complexity is isolated to the new gem binary classes.
+- **Storage requirements** - Platform-specific gems will increase repository storage needs on rubygems.org.
+- **Build matrix explosion** - Gem authors targeting multiple platforms need to build many more gem variants.
+
+# Rationale and Alternatives
+
+The gem binary format was chosen because:
+
+- It's a proven solution in the Python ecosystem with years of real-world usage
```

Comment:
Should we add a note that some people https://wheelnext.dev/ think that the current wheel have some problems?